# Name: Gengcong Yan
# Student Number: 1009903

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.metrics.pairwise import euclidean_distances

D:\Anaconda\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
D:\Anaconda\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
D:\Anaconda\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


# Task 1

In [10]:
# prepare data

ratdata= pd.read_csv("ratdataNormChecked.csv")

ratdata_ind=ratdata[["liverind", "heartind", "appind", "batind", "tailind", "ADWBind", "gonind","BMI"]]

ratdata_ind

liverind  heartind    appind    batind   tailind   ADWBind    gonind  \
0    0.036364  0.005091  0.013636  0.001864  0.714286  0.454545  0.000000   
1    0.010440  0.004396  0.008791  0.000742  0.666667  0.302198  1.648659   
2    0.051200  0.006133  0.020800  0.001467  0.904762  0.653333  0.000000   
3    0.047333  0.005556  0.105111  0.002178  0.869565  0.260000  2.653242   
4    0.052558  0.005349  0.021628  0.001884  0.720000  0.500000  0.000000   
..        ...       ...       ...       ...       ...       ...       ...   
556  0.067738  0.003507  0.013552  0.002602  0.808163  0.372172  1.742219   
557  0.078575  0.003667  0.014667  0.001035  0.777273  0.381090  1.064711   
558  0.079303  0.003918  0.014760  0.001337  0.760870  0.469952  0.993252   
559  0.080693  0.004286  0.008803  0.001592  0.714286  0.380252  1.512927   
560  0.085251  0.003682  0.016144  0.001218  0.816327  0.517429  0.405465   

          BMI  
0    0.199546  
1    0.478632  
2    0.340136  
3    0.340265  
4    0.275200  
..        ...  
556  0.736360  
557  0.788843  
558  0.786389  
559  0.793003  
560  0.764681  

[561 rows x 8 columns]

In [15]:
dis=euclidean_distances(ratdata_ind, ratdata_ind)
dis=pd.DataFrame(dis)
dis=dis.round(3)

dis

0      1      2      3      4      5      6      7      8      9    \
0    0.000  1.680  0.310  2.670  0.090  0.195  0.293  0.218  0.203  0.220   
1    1.680  0.000  1.709  1.040  1.674  1.689  1.560  1.592  1.675  1.693   
2    0.310  1.709  0.000  2.684  0.249  0.118  0.164  0.247  0.222  0.184   
3    2.670  1.040  2.684  0.000  2.670  2.674  2.549  2.575  2.661  2.686   
4    0.090  1.674  0.249  2.670  0.000  0.133  0.227  0.199  0.174  0.147   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
556  1.828  0.322  1.811  1.006  1.809  1.807  1.668  1.714  1.794  1.815   
557  1.221  0.678  1.194  1.660  1.190  1.188  1.047  1.098  1.169  1.196   
558  1.156  0.752  1.114  1.738  1.119  1.112  0.967  1.035  1.104  1.114   
559  1.627  0.361  1.614  1.247  1.604  1.606  1.466  1.518  1.595  1.609   
560  0.707  1.305  0.610  2.305  0.644  0.618  0.485  0.571  0.611  0.627   

     ...    551    552    553    554    555    556    557    558    559    560  
0    ...  0.746  0.925  1.301  0.712  0.522  1.828  1.221  1.156  1.627  0.707  
1    ...  1.744  0.946  0.518  1.721  1.688  0.322  0.678  0.752  0.361  1.305  
2    ...  0.590  0.892  1.279  0.587  0.408  1.811  1.194  1.114  1.614  0.610  
3    ...  2.745  1.948  1.497  2.720  2.699  1.006  1.660  1.738  1.247  2.305  
4    ...  0.661  0.887  1.277  0.638  0.439  1.809  1.190  1.119  1.604  0.644  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
556  ...  1.772  0.991  0.541  1.753  1.754  0.000  0.680  0.759  0.255  1.345  
557  ...  1.104  0.320  0.187  1.077  1.083  0.680  0.000  0.115  0.453  0.675  
558  ...  1.017  0.266  0.254  1.005  1.001  0.759  0.115  0.000  0.529  0.593  
559  ...  1.542  0.769  0.344  1.526  1.527  0.255  0.453  0.529  0.000  1.121  
560  ...  0.452  0.372  0.813  0.432  0.416  1.345  0.675  0.593  1.121  0.000  

[561 rows x 561 columns]

In [33]:
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.cluster import AgglomerativeClustering
from sklearn.decomposition import PCA


In [32]:
#linkage{‘ward’, ‘complete’, ‘average’, ‘single’}
link="single"

for numc in range(2,9):

    cluster = AgglomerativeClustering(n_clusters=numc, affinity='euclidean', linkage=link)  
    predicted_labels=cluster.fit_predict(ratdata_ind)
    # print("Predicted_labels:\n",predicted_labels)

    silhouette_avg = silhouette_score(ratdata_ind, predicted_labels)
    print("For n_clusters ={}, Linkage method={} The average silhouette_score is :{}".format(numc,link,silhouette_avg))

For n_clusters =2, Linkage method=single The average silhouette_score is :0.6571935462893095
For n_clusters =3, Linkage method=single The average silhouette_score is :0.422599240753836
For n_clusters =4, Linkage method=single The average silhouette_score is :-0.13713909132851268
For n_clusters =5, Linkage method=single The average silhouette_score is :-0.372639633794976
For n_clusters =6, Linkage method=single The average silhouette_score is :-0.3936197737503275
For n_clusters =7, Linkage method=single The average silhouette_score is :-0.407219449911702
For n_clusters =8, Linkage method=single The average silhouette_score is :-0.4617480200773739


# Original data
# Ward      K=3     SH=0.546
# Single    K=2     SH=0.657
# Complete  K=2     SH=0.597
# Average   K=2     SH=0.597

In [38]:
pca=PCA(n_components=1)
pca.fit(ratdata_ind)
# print(pca.explained_variance_ratio_)
data_pca=pca.fit_transform(ratdata_ind)
# data_pca

In [42]:
# Test in Data after PCA

#linkage{‘ward’, ‘complete’, ‘average’, ‘single’}
link="ward"

for numc in range(2,9):

    cluster = AgglomerativeClustering(n_clusters=numc, affinity='euclidean', linkage=link)  
    predicted_labels=cluster.fit_predict(data_pca)
    # print("Predicted_labels:\n",predicted_labels)

    silhouette_avg = silhouette_score(data_pca, predicted_labels)
    print("For n_clusters ={}, Linkage method={} The average silhouette_score is :{}".format(numc,link,silhouette_avg))

For n_clusters =2, Linkage method=ward The average silhouette_score is :0.5866555182619777
For n_clusters =3, Linkage method=ward The average silhouette_score is :0.6201023755102822
For n_clusters =4, Linkage method=ward The average silhouette_score is :0.5833175661083302
For n_clusters =5, Linkage method=ward The average silhouette_score is :0.5990347509737514
For n_clusters =6, Linkage method=ward The average silhouette_score is :0.5995483381988482
For n_clusters =7, Linkage method=ward The average silhouette_score is :0.6015072985915845
For n_clusters =8, Linkage method=ward The average silhouette_score is :0.6012152911910102


# Data after PCA
# Ward      K=3     SH=0.620
# Single    K=3     SH=0.675
# Complete  K=2     SH=0.622
# Average   K=3     SH=0.605

# c)
## single 2 0.657 in a)  size different  index 322 outlier
## single 3 0.675 in b)  size different  246 258 322
## feature gonind are wrong in some samples, for which can not exceed 1.


In [43]:
#linkage{‘ward’, ‘complete’, ‘average’, ‘single’}
link="single"

# a
cluster = AgglomerativeClustering(n_clusters=2, affinity='euclidean', linkage=link)  
predicted_labels_a=cluster.fit_predict(ratdata_ind)

cluster = AgglomerativeClustering(n_clusters=3, affinity='euclidean', linkage=link)  
predicted_labels_b=cluster.fit_predict(data_pca)

print("Predicted_labels_a:\n",predicted_labels_a)
print("Predicted_labels_b:\n",predicted_labels_b)

Predicted_labels_a:
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0

In [53]:
# a
indexs=[]
for i,x in enumerate(predicted_labels_a):
    if x!=0:
        print("index:",i)
        indexs.append(i)

ratdata_ind.iloc[indexs]

index: 322


liverind  heartind    appind    batind   tailind  ADWBind    gonind  \
322  0.033082  0.002518  0.012659  0.002056  0.860465      0.2  3.945264   

          BMI  
322  0.919416

In [54]:
# b
indexs=[]
for i,x in enumerate(predicted_labels_b):
    if x!=0:
        print("index:",i)
        indexs.append(i)

ratdata_ind.iloc[indexs]

index: 246
index: 258
index: 322


liverind  heartind    appind    batind   tailind   ADWBind    gonind  \
246  0.028558  0.002270  0.018652  0.002300  0.827273  0.125296  3.562182   
258  0.027572  0.002428  0.021788  0.002181  0.791111  0.171082  3.449988   
322  0.033082  0.002518  0.012659  0.002056  0.860465  0.200000  3.945264   

          BMI  
246  0.873967  
258  0.894815  
322  0.919416